### PR TITLE
C++ struct writer

### LINKED ISSUE #137: Implement C++ structs
Implement C++ code gen for structs

### LINKED ISSUE #138: Test C++ structs
Test C++ code gen for structs

### PR DESCRIPTION
**Pull request by tiffany1618**
Closes #137
Closes #138 

Added
- `StructCppWriter`: Writes C++ for FPP struct definitions
- Unit tests for C++ struct writer

Changes from Python XML struct autocoder:
- Add const and non-const getters for arrays and serializable type members, excluding enum types, which are passed by value (resolves issue of getting and then setting a struct member; e.g. a pattern like `s.get_a().set_b()`, where `s` is a struct containing a struct type member `a`, which in turn contains another member `b`)
- Pass array type members by C++ reference instead of by pointer and size (simpler, enforces checking of array length)
- Remove pointer copy constructor (reference copy constructor is sufficient)

### CODE REVIEW

**Comment by tiffany1618** [comment]
fpp-to-xml gives an error for empty structs, but fpp-to-cpp currently doesn't. Do we want fpp-to-cpp to also give an error?

**Comment by bocchino** [comment]
> fpp-to-xml gives an error for empty structs, but fpp-to-cpp currently doesn't. Do we want fpp-to-cpp to also give an error?

No. Empty structs are allowed in FPP and in C++. The error in fpp-to-xml exists because the XML schema does not support them. So this is a case where we are implementing part of the FPP spec that is not implemented when going through XML.

File: compiler/lib/src/main/scala/codegen/CppWriter/StructCppWriter.scala [hunk 1]
```diff
@@ -0,0 +1,775 @@
...
+          CppDoc.Lines.Both
+        )
+      ),
+      CppDoc.Class.Member.Constructor(
+        CppDoc.Class.Constructor(
+          Some("Constructor (default value)"),
+          Nil,
+          "Serializable()" :: nonArrayMemberNames.map(n => {
+            val defVal = defaultValues(n)
+            // Parse non-literal default values, e.g. T() for abstract type T
```
Comment (bocchino) [inline compiler/lib/src/main/scala/codegen/CppWriter/StructCppWriter.scala:236]: Why do we need this special-case handling? If we need it, let's not parse string output. Let's match against FPP semantic concepts, e.g., types.

**Review by bocchino — COMMENTED** [review]
Looks great! I have one comment so far. I'll schedule a review to go over the details.

**Comment by bocchino** [comment]
Also, same comment as on #144 -- please re-run all unit tests and update them as necessary.

File: compiler/lib/src/main/scala/codegen/CppWriter/StructCppWriter.scala (hunk 1 above)
Comment (bocchino) [reply compiler/lib/src/main/scala/codegen/CppWriter/StructCppWriter.scala]: Looks good!

**Review by bocchino — APPROVED** [review]
Looks good! I confirmed that the unit tests pass.

**Pull request merged**